Library/Homebrew/cmd/tests.rb

```diff
@@ -36,18 +24,55 @@ def tests
         args << "TESTOPTS=--name=test_#{test_method}" if test_method
       end
       args += ARGV.named.select { |v| v[/^TEST(OPTS)?=/] }
-      system "bundle", "exec", "rake", "test", *args
 
-      Homebrew.failed = !$?.success?
+      begin
+        sanitize_env!
+
+        # Since we "borrow" HOME for duration of tests bundler
+        # needs fetching/installing every run.
+        Homebrew.install_gem_setup_path! "bundler"
+        unless quiet_system("bundle", "check")
+          system "bundle", "install", "--path", "vendor/bundle"
+        end
+
+        system "bundle", "exec", "rake", "test", *args
+        Homebrew.failed = !$?.success?
 
-      if (fs_leak_log = HOMEBREW_LIBRARY/"Homebrew/test/fs_leak_log").file?
-        fs_leak_log_content = fs_leak_log.read
-        unless fs_leak_log_content.empty?
-          opoo "File leak is detected"
-          puts fs_leak_log_content
-          Homebrew.failed = true
+        if (fs_leak_log = HOMEBREW_LIBRARY/"Homebrew/test/fs_leak_log").file?
+          fs_leak_log_content = fs_leak_log.read
+          unless fs_leak_log_content.empty?
+            opoo "File leak is detected"
+            puts fs_leak_log_content
+            Homebrew.failed = true
+          end
         end
+      ensure
+        ENV["HOME"] = @old_home
+        # For the sake of safety, ensure @env_home actually resides in
+        # HOMEBREW_TEMP before trying to remove this directory.
+        FileUtils.rm_r @env_home if @env_home.dirname == HOMEBREW_TEMP
```

Comment:
I don't feel super strongly about this, but I'm not sure if `@old_name` and `@env_home` are clear enough. How about `@true_home` and `@test_home`?